src/context/trace_context.rs

```diff
@@ -30,39 +30,86 @@ use crate::{
 use std::{
     fmt::Formatter,
     mem::take,
-    sync::{
-        atomic::{AtomicI32, Ordering},
-        Arc, Mutex, RwLock, Weak,
-    },
+    sync::{Arc, RwLock},
 };
 
-struct Inner {
-    trace_id: RwLock<String>,
-    trace_segment_id: String,
-    service: String,
-    service_instance: String,
-    next_span_id: AtomicI32,
-    spans: Mutex<Vec<SpanObject>>,
-    active_span_stack: Mutex<Vec<SpanObject>>,
-    segment_link: Option<PropagationContext>,
-    primary_endpoint_name: Mutex<String>,
+#[derive(Default)]
+pub(crate) struct SpanStack {
+    pub(crate) finialized: RwLock<Vec<SpanObject>>,
+    pub(crate) active: RwLock<Vec<SpanObject>>,
+}
+
+impl SpanStack {
+    pub(crate) fn with_finialized<T>(&self, f: impl FnOnce(&Vec<SpanObject>) -> T) -> T {
+        f(&self.finialized.try_read().expect(LOCK_MSG))
+    }
+
+    pub(crate) fn with_finialized_mut<T>(&self, f: impl FnOnce(&mut Vec<SpanObject>) -> T) -> T {
+        f(&mut *self.finialized.try_write().expect(LOCK_MSG))
+    }
+
+    pub(crate) fn with_active<T>(&self, f: impl FnOnce(&Vec<SpanObject>) -> T) -> T {
+        f(&*self.active.try_read().expect(LOCK_MSG))
+    }
+
+    pub(crate) fn with_active_mut<T>(&self, f: impl FnOnce(&mut Vec<SpanObject>) -> T) -> T {
+        f(&mut *self.active.try_write().expect(LOCK_MSG))
+    }
+
+    fn pop_active(&self, index: usize) -> Option<SpanObject> {
+        self.with_active_mut(|stack| {
+            if stack.len() > index + 1 {
+                None
+            } else {
+                stack.pop()
+            }
+        })
+    }
+
+    /// Close span. We can't use closed span after finalize called.
+    pub(crate) fn finalize_span(&self, index: usize) {
+        let span = self.pop_active(index);
+        if let Some(mut span) = span {
+            span.end_time = fetch_time(TimePeriod::End);
+            self.with_finialized_mut(|spans| spans.push(span));
+        } else {
+            panic!("Finalize span isn't the active span");
+        }
+    }
 }
 
 #[must_use = "call `create_entry_span` after `TracingContext` created."]
 pub struct TracingContext {
-    inner: Arc<Inner>,
+    trace_id: String,
+    trace_segment_id: String,
+    service: String,
+    service_instance: String,
+    next_span_id: i32,
+    span_stack: Arc<SpanStack>,
+    segment_link: Option<PropagationContext>,
```

Comment:
Why is a segment link? Segment reference should be inside a span, and it should be multiple references.

Reply:
This field existed before I was involved in this project. 😂

Reply:
Let's check why it exists. and try to remove it?

Reply:
Ok, actually I also think this field is not very right.